src/App.tsx

```diff
@@ -54,48 +69,100 @@ export default function App() {
-
-    const [bbox, setBbox] = useState<Bbox>([-180, -90, 180, 90])
     // todo: maybe combine these effects into one? see discussion in #77
     useEffect(() => {
         // make sure the path bbox and the path details bbox take precedence over the info bbox
-        if (!route.selectedPath.bbox && !pathDetails.pathDetailBbox) setBbox(info.bbox)
+        if (!route.selectedPath.bbox && !pathDetails.pathDetailBbox)
+            Dispatcher.dispatch(new SetViewportToBbox(info.bbox))
     }, [info])
     useEffect(() => {
         // make sure the path details bbox takes precedence over the route bbox
-        if (route.selectedPath.bbox && !pathDetails.pathDetailBbox) setBbox(route.selectedPath.bbox)
+        if (route.selectedPath.bbox && !pathDetails.pathDetailBbox)
+            Dispatcher.dispatch(new SetViewportToBbox(route.selectedPath.bbox))
     }, [route])
     useEffect(() => {
         // make sure the path details bbox takes precedence over the path and info bboxes
-        if (pathDetails.pathDetailBbox) setBbox(pathDetails.pathDetailBbox)
-        else if (route.selectedPath.bbox) setBbox(route.selectedPath.bbox)
+        if (pathDetails.pathDetailBbox) Dispatcher.dispatch(new SetViewportToBbox(pathDetails.pathDetailBbox))
+        else if (route.selectedPath.bbox) Dispatcher.dispatch(new SetViewportToBbox(route.selectedPath.bbox))
     }, [pathDetails])
 
+    const [popupCoordinate, setPopupCoordinate] = useState<Coordinate | null>(null)
+    const theMapLayers: { [key: string]: MapLayer } = {
+        'context-menu-layer': {
+            id: 'context-menu-layer',
+            interactiveLayerIds: [],
+            onClick: () => {},
+            layer: (
+                <ContextMenuLayer
+                    queryPoints={query.queryPoints}
+                    popupCoordinate={popupCoordinate}
+                    setPopupCoordinate={setPopupCoordinate}
+                />
+            ),
+        },
+        'query-points-layer': {
+            id: 'query-points-layer',
+            interactiveLayerIds: [],
+            onClick: () => {},
+            layer: <QueryPointsLayer queryPoints={query.queryPoints} />,
+        },
+        'paths-layer': {
+            id: 'paths-layer',
+            interactiveLayerIds: getInteractiveLayerIds(route.selectedPath, route.routingResult.paths),
+            onClick: feature => {
+                // select an alternative path if clicked
+                if (feature.layer.id === pathsLayerKey) {
+                    const index = feature.properties!.index
+                    const path = getCurrentPaths(route.selectedPath, route.routingResult.paths).find(
+                        indexPath => indexPath.index === index
+                    )
+                    Dispatcher.dispatch(new SetSelectedPath(path!.path))
+                }
+            },
+            layer: <PathsLayer selectedPath={route.selectedPath} paths={route.routingResult.paths} />,
+        },
```

Comment:
@Janekdererste maybe you can help me here? Basically I would like to move all this into `PathsLayer.tsx` for example, so here in `App.tsx` it would maybe simply look like this: 

```javascript
const mapLayers = [
  new ContextMenuLayer(popupCoordinate),
  new QueryPointLayer(query.queryPoints),
  new PathsLayer(route.selectedPath, route.routingResult.paths),
  new PathDetailsLayer(pathDetails),
]
mapLayers.forEach(layer => {
  Dispatcher.dispatch(new PutMapLayerAction(layer))
})
```

But I could not figure out how to move the JSX out from here. Note that I already set up the `MapLayerStore` but not using it so far.

Reply:
My goal was that adding a layer would be a matter of adding a file that specifies the layer and calling `Dispatch.dispatch(new PutMapLayerAction(layer))` somewhere. The latter is probably done only when the app loads, but in principle we might add/remove layers interactively?! We need the `interactiveLayerIds` and `onClick` propertys for every such layer, because the Mapbox/React wrapper needs to pass the user events to all our layers.

Reply:
Sorry I went back and forth here a bit and now I actually minimized the changes and  think this should be included in #87